packages/core/src/ui/UIRenderer.ts

```diff
@@ -0,0 +1,193 @@
+    return this._rayCastPadding;
+  }
+
+  set rayCastPadding(value: Vector4) {
+    if (this._rayCastPadding !== value) {
+      this._rayCastPadding.copyFrom(value);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  constructor(entity: Entity) {
+    super(entity);
+    this._componentType = ComponentType.UIRenderer;
+  }
+
+  /**
+   * @internal
+   */
+  override _updateTransformShaderData(context: RenderContext, onlyMVP: boolean, batched: boolean): void {
+    //@todo: Always update world positions to buffer, should opt
+    // super._updateTransformShaderData(context, onlyMVP, true);
+    const worldMatrix = this.entity.transform.worldMatrix;
+    if (onlyMVP) {
+      this._updateProjectionRelatedShaderData(context, worldMatrix, true);
+    } else {
+      this._updateWorldViewRelatedShaderData(context, worldMatrix, true);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  override _prepareRender(context: RenderContext): void {
+    // Update once per frame per renderer, not influenced by batched
+    if (this._renderFrameCount !== this.engine.time.frameCount) {
+      this._update(context);
+    }
+
+    this._render(context);
+
+    // union camera global macro and renderer macro.
+    ShaderMacroCollection.unionCollection(
+      context.camera._globalShaderMacro,
+      this.shaderData._macroCollection,
+      this._globalShaderMacro
+    );
+  }
+
+  /**
+   * @internal
+   */
+  override _onEnableInScene(): void {
+    const componentsManager = this.scene._componentsManager;
+    if (this._overrideUpdate) {
+      componentsManager.addOnUpdateRenderers(this);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  override _onDisableInScene(): void {
+    const componentsManager = this.scene._componentsManager;
+    if (this._overrideUpdate) {
+      componentsManager.removeOnUpdateRenderers(this);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  _setGroup(group: CanvasGroup): void {
+    this._group = group;
+    const alpha = group ? group._globalAlpha : 1;
+    if (this._alpha !== alpha) {
+      this._alpha = alpha;
+      this._dirtyUpdateFlag |= UIRendererUpdateFlags.Alpha;
+    }
+  }
+
+  /**
+   * @internal
+   */
+  _getChunkManager(): PrimitiveChunkManager {
+    return this.engine._batcherManager.primitiveChunkManagerUI;
+  }
+
+  /** @internal */
+  _raycast(ray: Ray, out: HitResult, distance: number = Number.MAX_SAFE_INTEGER): boolean {
+    const entity = this._entity;
+    const plane = UIRenderer._tempPlane;
+    const transform = entity.transform;
+    const normal = plane.normal.copyFrom(transform.worldForward);
+    plane.distance = -Vector3.dot(normal, transform.worldPosition);
+    ray.intersectPlane(plane);
+    const curDistance = ray.intersectPlane(plane);
+    if (curDistance >= 0 && curDistance < distance) {
+      const hitPointWorld = ray.getPoint(curDistance, UIRenderer._tempVec30);
+      const worldMatrixInv = UIRenderer._tempMat;
+      Matrix.invert(this.entity.transform.worldMatrix, worldMatrixInv);
```

Comment:
As comment before, we can destructure worldMatrix , here we can use it directly.